### PR TITLE
Make IPv6 Great Again

### PR DESCRIPTION
**Pull request by tgraf**
This makes IPv6 a proper citizen by:
- Allowing an endpoint to have only an IPv6 address if network.enableIPv6 is set.
- Enabling allocation of an IPv6 address if netlabel.EnableIPv6 is set even without the presence of a network.ipamV6Conf.

Existing IPv4 behaviour is kept intact: Unless network.enableIPv6 is explicitly set by the user, an IPv4 configuration failure is considered fatal even in the presence of a working IPv6 configuration.

### CODE REVIEW

File: network.go [hunk 1]
```diff
@@ -931,7 +931,7 @@ func (n *network) ipamAllocateVersion(ipVer int, ipam ipamapi.Ipam) error {
 	}
 
 	if len(*cfgList) == 0 {
-		if ipVer == 6 {
+		if ipVer == 6 && !n.enableIPv6 {
```
Comment (aboch) [inline network.go]: I think this check should instead take into account whether the ipam driver supports auto-allocation of ipv6 pools.
Now that `ipamapi.Capability` is in master, I'd suggest to add a `SupportsAutoIPv6` bool to the ipam driver capability.

In fact the reason why this function was bailing out in case of empty `ipamV6Config` was just because the in-built ipam driver does not have the capability of self assigning an ipv6 pool from some predefined list.

File: ipams/remote/remote.go [hunk 2]
```diff
@@ -93,6 +93,8 @@ func (a *allocator) RequestAddress(poolID string, address net.IP, options map[st
 	}
 	if res.Address != "" {
 		retAddress, err = types.ParseCIDR(res.Address)
+	} else {
+		return nil, nil, ipamapi.ErrNoAvailableIPs
```
Comment (aboch) [inline ipams/remote/remote.go]: The allocation errors (no available ips, ip already allocated, ...) should be returned by the remote driver in `RequestAddress()` response `RequestAddressResponse.Response.Error`

If the generic call fails `a.call` will return the request's error (all responses embed `api.Response`).

File: endpoint.go [hunk 3]
```diff
@@ -767,13 +767,15 @@ func (ep *endpoint) assignAddress(assignIPv4, assignIPv6 bool) error {
...
 	if assignIPv4 {
-		if err = ep.assignAddressVersion(4, ipam); err != nil {
+		if err = ep.assignAddressVersion(4, ipam); err != nil && !ep.network.enableIPv6 {
 			return err
 		}
 	}
 
 	if assignIPv6 {
-		err = ep.assignAddressVersion(6, ipam)
+		if err = ep.assignAddressVersion(6, ipam); err != nil && ep.network.enableIPv6 {
```
Comment (aboch) [inline endpoint.go]: If we allow the auto-ipv6 allocation via capability, this check does not need to be modified.

Also, in case ipv6 is enabled and the user specifies an IPv6 subnet (no auto allocation), this change would hide a legitimate ipv6 address allocation failure, ehich is not correct.

File: endpoint.go [hunk 4]
```diff
@@ -837,8 +839,10 @@ func (ep *endpoint) releaseAddress() {
 		log.Warnf("Failed to retrieve ipam driver to release interface address on delete of endpoint %s (%s): %v", ep.Name(), ep.ID(), err)
 		return
 	}
-	if err := ipam.ReleaseAddress(ep.iface.v4PoolID, ep.iface.addr.IP); err != nil {
-		log.Warnf("Failed to release ip address %s on delete of endpoint %s (%s): %v", ep.iface.addr.IP, ep.Name(), ep.ID(), err)
+	if ep.iface.addr != nil {
```
Comment (aboch) [inline endpoint.go]: This is needed. Once we support the `ipv6-only` scenario (via appropriate flag), this situation may occur.

File: network.go [hunk 5]
```diff
@@ -931,7 +931,7 @@ func (n *network) ipamAllocateVersion(ipVer int, ipam ipamapi.Ipam) error {
...
 			return nil
 		}
```
Comment (aboch) [inline network.go]: In order to support the `--ipv6-only` scenario, we should bail out here in case of v4:

```
if ipVer == 4 && n.ipv6Only {
    return nil
}
```

File: endpoint.go [hunk 6]
```diff
@@ -767,13 +767,15 @@ func (ep *endpoint) assignAddress(assignIPv4, assignIPv6 bool) error {
 	}
 
 	if assignIPv4 {
-		if err = ep.assignAddressVersion(4, ipam); err != nil {
+		if err = ep.assignAddressVersion(4, ipam); err != nil && !ep.network.enableIPv6 {
```
Comment (aboch) [inline endpoint.go]: I think this change is needed only to get the things work in case we overload `enableIPv6` with the `allocate ipv6 only` message.

**Comment by aboch** [comment]
@tgraf 
Thanks for the change and help, we indeed need help to fill in some ipv6 missing parts in libnetwork.

Put few comments. In summary, I think we should not overload the `enableIPv6` flag with the `allocate only ipV6` meaning.

I see this PR is trying to achieve at least two distinct things: 
- Allow auto allocation of IPv6 pools
- Allow ipv6 only networks

For the first one I'd suggest to let the ipam driver express he can support it (now we have ipam capability), for the second one I think this has to come from user first class config, maybe an `--ipv6only` UI flag, as I was also saying in https://github.com/docker/docker/issues/18607#issuecomment-164498509

WDYT ?

File: network.go (hunk 1 above)
Comment (tgraf) [inline network.go]: That's fine as well, in fact I would even prefer if the user does not have to explicitly request it.

File: ipams/remote/remote.go (hunk 2 above)
Comment (tgraf) [inline ipams/remote/remote.go]: Right but that wouldn't handle a misbehaving driver. A driver returning an empty address will cause `RequestAddr()` to return `nil` because of the conditional assignment of `redAddress`. I'm happy to fix this in another way but I don't think we should depend on a well behaving driver here.

File: endpoint.go (hunk 3 above)
Comment (tgraf) [inline endpoint.go]: I just noticed that this change is not needed at all because the error is returned a couple of lines down anyway. I will drop this.

**Comment by tgraf** [comment]
@aboch What I had in mind is to add a `--ipv6[=true]` option to `network create` which sets `netlabel.EnableIPv6` which would indicate that IPv6 should be treated the same way as IPv4, i.e. not require a successful IPv4 address configuration in order for the networking setup to succeed. I think we have three scenarios:
 1. Trigger allocation of IPv6 but still fail if IPv4 failed
 2. Trigger allocation of IPv6 and succeed if either IPv6 or IPv4 have succeeded
 3. Enforce an IPv6 only configuration, no IPv4 allocation at all

The IPAM capability allows to do (1) automatically but I doubt we can go to (2) without an additional user option such as `--ipv6` as this might break current expectations. Users might rely on this current failure behavior. (2) reflects what this PR reflects so far and (3) would require the `--ipv6-only` flag that you mentioned.

File: ipams/remote/remote.go (hunk 2 above)
Comment (aboch) [inline ipams/remote/remote.go]: Ah ok, if it's for misbehaving driver. Given the driver is not following the protocol, let's return an error indicating that, rather than treating it as a "no available IPs" one.

**Comment by aboch** [comment]
@tgraf 
Thanks for the clarification.

I am not sure about the option 2 you mentioned. Considering the behavior would be controlled by a user specified option, how would it be useful to create a network which may get either an ipv4 or an ipv6 pool without knowing it a priori ?

In a scenario where the ipam driver and network driver are implemented by the same entity, I am guessing why this would seem to be needed: It would give freedom to the network driver to decide whether to assign ipv4 or ipv6 pool for a specific network, without libnetwork complaining about it.

But this would go against the reason for which the IPAM was added into libnetwork in first place, which is to give the docker user control over the ip address management for the containers/networks. The assigned ipam behavior should be deterministic.

Probably we should allow the ipam driver also express its capability in supporting only v4, only v6 or both during registration. This way libnetwork will handle the ipam request accordingly and, as an example, won't ask for ipv4 pool if ipam driver does not or does not want to support it. This would be more explicit and confined to the iapm driver.
At the end user chooses the ipam driver, he is expected to know its behavior/capabilities.

Regarding `netlabel.EnableIPv6`, its current use has it carrying the semantic  `enable ipv6 besides ipv4`, in order to satisfy the existing `docker daemon` `--ipv6` option.
(There is also a proposal to add it to the `docker network` ui as first class option, currently in there it can only be passed as driver option).

In other words, ipv6 is currently considered optional and complementary to ipv4.

**Comment by tgraf** [comment]
@aboch 
I've rebased the changes and picked up your suggestion to use a new IPAM capability. The new capability allows the IPAM driver to indicate whether to use IPv4, IPv6, or both. If the capability is not present, the behaviour is as before. This does not require any additional user input at all and the entire decision is delegated to the IPAM.

Let me know what you think and I'll add a couple of test cases as well.

**Comment by tgraf** [comment]
Rebased. Ping @mavenugo @aboch.

**Comment by mavenugo** [comment]
@tgraf apologies on the delay. we will get to this shortly

**Comment by tgraf** [comment]
Casual ping @mavenugo

**Comment by mavenugo** [comment]
@tgraf we couldn't prioritize this for the 1.10 release. And there are known ipv6 limitations which we are attempting to solve in 1.11 & beyond. Please let us know If you are interested in contributing to ipv6 fixes and bring it in par with the ipv4 handling.
As far as this PR goes, we will get to this once the top priority issues for 1.10 is taken care of.

**Comment by tgraf** [comment]
@mavenugo Sure. What issues are you referring to?

**Comment by aidanhs** [comment]
> What I had in mind is to add a --ipv6[=true] option to network create which sets netlabel.EnableIPv6

FYI docker/docker#17513 will do exactly this. The exact detail of what that flag will *mean* is (of course) up to libnetwork.

**Comment by tgraf** [comment]
@aidanhs This looked like the right thing to do first but I don't see a reason anymore why the user should be required to opt into IPv6. Whether to use IPv6 or IPv4 or both depends on the infrastructure and should be handled at the driver level. At least that possibility should be given.

**Comment by aidanhs** [comment]
Sure, maybe the default will change, but people may still want to opt out so the flag is still necessary.

**Comment by tgraf** [comment]
@mavenugo Ping, any interest in going forward with this?

File: docs/ipam.md [hunk 7]
```diff
@@ -263,7 +264,10 @@ As of now libnetwork accepts the following capabilities:
 It is a boolean value which tells libnetwork whether the ipam driver needs to know the interface MAC address in order to properly process the `RequestAddress()` call.
 If true, on `CreateEndpoint()` request, libnetwork will generate a random MAC address for the endpoint (if an explicit MAC address was not already provided by the user) and pass it to `RequestAddress()` when requesting the IP address inside the options map. The key will be the `netlabel.MacAddress` constant: `"com.docker.network.endpoint.macaddress"`.
 
+### SupportsAutoIPv6
+
+This boolean tells libnetwork that the ipam driver is capable of deciding whether an IPv4 address, IPv6 address, or both should be assigned to the container by returning either an address or an empty string in response to `RequestAddress()`.
```
Comment (aboch) [inline docs/ipam.md]: I would not overload this flag with two meaning. The name seems clear, it tells whether the IPAM driver can come up with a IPv6 pool if user did not configure one.

File: endpoint.go [hunk 8]
```diff
@@ -944,6 +944,10 @@ func (ep *endpoint) assignAddressVersion(ipVer int, ipam ipamapi.Ipam) error {
 			ep.Unlock()
 			return nil
 		}
+		if err == ipamapi.ErrNoIPReturned && ipam.capability.SupportsAutoIPv6 {
```
Comment (aboch) [inline endpoint.go]: I do not think this function needs this change.
If the IPAM driver, say, does not handle IPv4, then I would expect `len(ipInfo) == 0` above, and this function can exit, regardless of `ipVer` (need to change this).

Because I would expect `ipamAllocateVersion()` to not initialize `network.ipamV4Info` if the IPAM driver does not or does not want to support autoallocation of IPv4 pools.

File: ipamapi/contract.go [hunk 9]
```diff
@@ -50,6 +50,7 @@ var (
 	ErrIPOutOfRange        = types.BadRequestErrorf("Requested address is out of range")
 	ErrPoolOverlap         = types.ForbiddenErrorf("Pool overlaps with other one on this address space")
 	ErrBadPool             = types.BadRequestErrorf("Address space does not contain specified address pool")
+	ErrNoIPReturned        = types.NoServiceErrorf("No address returned")
```
Comment (aboch) [inline ipamapi/contract.go]: Looking at this new error, it makes me think we may want to follow the same idea but for the Pool, something like
`ErrNoPoolReturned`.
Once a a pool is not returned for a network, libnetwork would not even query the driver for an address, during create endpoint.

**Comment by aboch** [comment]
@tgraf Sorry for the back and forth.

Looking at your new diffs, I am wondering whether it would be better to follow the approach you are using for request address (driver to return a specific error) for pool request instead, and drop the new capability flag.
At first look, it seems we no longer need a new capability flag and it is better because I feel it is hard to come out with good and meaningful ones which will not need to change in future.

We would just need to error out in case user has specified a `--subnet` but ipam driver did return a service error.

Do you think it makes sense ?

Also, your changes would likely not make it for 1.11, code freeze is today, but we are committed to have them in 1.12.

Thanks

**Comment by jainvipin** [comment]
@aboch, @mavenugo - no milestone attached to it. I was wondering if this PR will be merged sometime soon?

**Comment by tgraf** [comment]
@aboch I addressed your feedback and reverted the initial suggestion of the capability. The proposal is now back to the original approach. I left out the gateway opt out decision for now. Let me know what you think.

**Comment by tgraf** [comment]
ping @nerdalert @mrjana @mavenugo

**Comment by tgraf** [comment]
ping @aboch

**Comment by tgraf** [comment]
@mrjana @aboch Any feedback on what is left to get this merged? It is blocking various people wanting to run IPv6 only containers.

**Comment by mrjana** [comment]
@tgraf The way we have to introduce ipv6 only networks is by introducing an `ipv6-only` option which is accessible to the user and then we can make necessary changes to say if the network is `ipv6-only` then we will skip ipv4 allocations/plumbing and just do an e2e ipv6 plumbing and we can also make sure the driver(any driver) can handle an `ipv6-only` network (or not in which case it returns an error). This ensures that the current behavior doesn't change and the new behavior is only present when `ipv6-only` option is provided by the user.

**Comment by tgraf** [comment]
@mrjana Isn't that option already there in the form of `--enable-ipv6`. It's an explicit opt-in already.

**Comment by pneumaticdeath** [comment]
Just a vote of support for this.  This (and #23090 in kubernetes) and blocking us from adopting k8s/docker as a container solution.

**Comment by RsrchBoy** [comment]
...and another +1

**Comment by choppsv1** [comment]
Really need ipv6 only *and* auto ipv6 allocation. Hard to do IPv6 development w/o these things.

**Comment by agowa338** [comment]
@mrjana @aboch @tgraf What's the current status of this pr?
We're still waiting for this in https://github.com/moby/moby/issues/32850
Is this pr dead and someone should fill a new one?

**Comment by telmich** [comment]
Very much interested in this, too. Especially because we need to turn IPv4 off in various network constellations nowadays

**Comment by Citrullin** [comment]
Would be also interested in this. We should all move to IPv6 only networks.

**Comment by murphy83** [comment]
Any Update on this? IPv6-only would save me a lot of hassle with exhausted Ipv4-Networks during VPN-Connections at some customers ...

**Comment by frebib** [comment]
It would be great to get this into 20.10. This problem alone is making me consider replacing Docker for something more forward thingking

**Comment by Robotic-Brain** [comment]
> It would be great to get this into 20.10. This problem alone is making me consider replacing Docker for something more forward thinking

Which would be?
I'm desperately searching for an alternative, which doesn't devolve into creating everything from scratch with nspawn!
Something which gives me full control over networking but understands the docker image format would be perfect.

**Comment by rpodgorny** [comment]
> > It would be great to get this into 20.10. This problem alone is making me consider replacing Docker for something more forward thinking
> 
> Which would be? I'm desperately searching for an alternative, which doesn't devolve into creating everything from scratch with nspawn! Something that gives me full control over networking but understands the docker image format would be perfect.

podman?

**Comment by Robotic-Brain** [comment]
> podman?

As far as i can tell podman is just as broken as docker is when it comes to IPv6-only networking because CNI also relies on deprecated NAT technology. (but I'll look into it)
